### PR TITLE
Initial support for func_plat entity

### PR DESCRIPTION
**Pull request by Mac-tf2**
Isn't exactly 1:1 with quake but it's a working elevator and it stops maps with func_plat from crashing. I can make it closer to quake elevators with a bunch more entity connections if needed but this seems fine for now.

### CODE REVIEW

File: BSPConvert.Lib/Source/EntityConverter.cs [hunk 1]
```diff
@@ -209,6 +213,80 @@ private void ConvertFuncStatic(Entity funcStatic)
...
+			var brushThickness = GetBrushThickness(entity);
+
+			if (float.TryParse(entity["height"], out var height))
+				moveDistance = height + brushThickness;
+			else if (float.TryParse(entity["lip"], out var lip))
+				moveDistance = -(lip - 2 - (brushThickness * 2));
+
+			if (string.IsNullOrEmpty(entity.Name))
+			{
+				entity.Name = $"plat{platIndex}";
```
Comment (Panzerhandschuh) [inline BSPConvert.Lib/Source/EntityConverter.cs]: Renaming this entity will make it so it cannot be found in `entityDict`. This might not be an issue if `func_plat` is never the target of another entity though.

File: BSPConvert.Lib/Source/EntityConverter.cs [hunk 2]
```diff
@@ -209,6 +213,80 @@ private void ConvertFuncStatic(Entity funcStatic)
...
+				CreatePlatTrigger(entity);
```
Comment (Panzerhandschuh) [inline BSPConvert.Lib/Source/EntityConverter.cs:229]: This trigger won't get created if the `func_plat` has an existing name. Is this intended?

File: BSPConvert.Lib/Source/EntityConverter.cs [hunk 3]
```diff
@@ -209,6 +213,80 @@ private void ConvertFuncStatic(Entity funcStatic)
...
 
+		private void ConvertFuncPlat(Entity entity)
+		{
+			var moveDistance = 0f;
+			var brushThickness = GetBrushThickness(entity);
+
+			if (float.TryParse(entity["height"], out var height))
+				moveDistance = height + brushThickness;
+			else if (float.TryParse(entity["lip"], out var lip))
+				moveDistance = -(lip - 2 - (brushThickness * 2));
```
Comment (Panzerhandschuh) [inline BSPConvert.Lib/Source/EntityConverter.cs]: What does the `- 2` mean? You should declare a constant with a name that describes what it is being used for.

File: BSPConvert.Lib/Source/EntityConverter.cs [hunk 4]
```diff
@@ -209,6 +213,80 @@ private void ConvertFuncStatic(Entity funcStatic)
...
+				param = null,
+				delay = 3,
+				fireOnce = -1
+			};
+			plat.connections.Add(connection2);
+		}
+
+		private float GetBrushThickness(Entity entity)
+		{
+			var modelIndexStr = entity["model"].Substring(1); // Removes * from model index
```
Comment (Panzerhandschuh) [inline BSPConvert.Lib/Source/EntityConverter.cs]: Can you get the index directly from the `entity.ModelNumber` property?

File: BSPConvert.Lib/Source/EntityConverter.cs [hunk 5]
```diff
@@ -209,6 +213,80 @@ private void ConvertFuncStatic(Entity funcStatic)
...
+			if (!int.TryParse(modelIndexStr, out var modelIndex))
+				return 0;
+			var model = q3Models[modelIndex];
```
Comment (Panzerhandschuh) [inline BSPConvert.Lib/Source/EntityConverter.cs]: (nit) I typically put a new line after single line if statements. I think it helps with readability a bit and makes it clear that this code is unrelated to the if statement.

```
if (!int.TryParse(modelIndexStr, out var modelIndex))
    return 0;

var model = q3Models[modelIndex];
```

File: BSPConvert.Lib/Source/EntityConverter.cs [hunk 6]
```diff
@@ -209,6 +213,80 @@ private void ConvertFuncStatic(Entity funcStatic)
...
+			};
+			trigger.connections.Add(connection);
+
+			var connection2 = new Entity.EntityConnection()
+			{
+				name = "onFullyClosed",
+				target = plat.Name,
+				action = "open",
+				param = null,
+				delay = 3,
```
Comment (Panzerhandschuh) [inline BSPConvert.Lib/Source/EntityConverter.cs]: Is this delay hard coded into Quake 3? It might be worth adding a comment here explaining that.

File: BSPConvert.Lib/Source/EntityConverter.cs (hunk 2 above)
Comment (Mac-tf2) [reply BSPConvert.Lib/Source/EntityConverter.cs:229]: I was going off the q3 source code, but I might be misinterpreting it.

``` 
	if ( !ent->targetname ) {
		SpawnPlatTrigger(ent);
```

Looks like the trigger is only made when the entity doesnt have a targetname.

File: BSPConvert.Lib/Source/EntityConverter.cs (hunk 6 above)
Comment (Mac-tf2) [reply BSPConvert.Lib/Source/EntityConverter.cs]: The platform never resets in q3 as long as a player is touching it. It resets 1 second after you step off. I can _probably_ recreate this with entity logic in the future but it would need a bunch more entity connections that will look pretty messy. This is just a placeholder value to make it work in the meantime. I can comment it as such.

File: BSPConvert.Lib/Source/EntityConverter.cs [hunk 7]
```diff
@@ -209,6 +212,77 @@ private void ConvertFuncStatic(Entity funcStatic)
...
+				param = null,
+				delay = 3, // placeholder value TODO: replicate actual func_plat behaviour
+				fireOnce = -1
+			};
+			plat.connections.Add(connection2);
+		}
+
+		private float GetBrushThickness(Entity entity)
+		{
+			var model = q3Models[entity.ModelNumber];
```
Comment (Panzerhandschuh) [inline BSPConvert.Lib/Source/EntityConverter.cs:281]: I think this can be simplified further to just `var model = entity.Model;`

Comment (Mac-tf2) [reply BSPConvert.Lib/Source/EntityConverter.cs:281]: entity.Model is just the model number in string form, only q3Models have model size data as far as I can tell.

Comment (Mac-tf2) [reply BSPConvert.Lib/Source/EntityConverter.cs:281]: image
